dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -57,42 +62,76 @@ public WorkerLocationPolicy(int numVirtualNodes) {
   public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWorkerInfos,
                                                    String fileId,
                                                    int count) {
-    if (blockWorkerInfos.size() == 0) {
+    if (blockWorkerInfos.isEmpty()) {
       return ImmutableList.of();
     }
     HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
     return HASH_PROVIDER.getMultiple(fileId, count);
   }
 
-  private static class ConsistentHashProvider {
+  static class ConsistentHashProvider {
     private static final HashFunction HASH_FUNCTION = murmur3_32_fixed();
-    private static final int MAX_ATTEMPTS = 100;
-    private List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
-    private NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
-
-    private volatile long mLastUpdatedTimestamp = 0L;
-
-    private final AtomicBoolean mNeedUpdate = new AtomicBoolean(false);
-
-    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 1000L;
+    private final int mMaxAttempts;
+    private final long mWorkerInfoUpdateIntervalNs;
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    // Lazily initialized, can only be null before the first call to refresh
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
+    // Must use System.nanoTime to ensure monotonic increment
+
+    public ConsistentHashProvider(int maxAttempts, long workerListTtlMs) {
+      mMaxAttempts = maxAttempts;
+      mWorkerInfoUpdateIntervalNs = workerListTtlMs * Constants.MS_NANO;
+    }
 
+    /**
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      Preconditions.checkArgument(!workerInfos.isEmpty(),
+          "cannot refresh hash provider with empty worker list");
+      maybeInitialize(workerInfos, numVirtualNodes);
+      // check if the worker list has expired
+      long lastUpdateTs = mLastUpdatedTimestamp.get();
+      long currentTs = System.nanoTime();
+      if (currentTs - lastUpdateTs > mWorkerInfoUpdateIntervalNs) {
+        // use CAS to only allow one thread to actually update the timestamp
+        boolean casUpdated = mLastUpdatedTimestamp.compareAndSet(lastUpdateTs, currentTs);
+        // thread safety is valid provided that build() takes less than
+        // WORKER_INFO_UPDATE_INTERVAL_NS, so that before next update the current update has been
+        // finished
+        if (casUpdated) {
```

Comment:
Extract a method?
```
if (shouldRebuildHashRingExclusively())  { // extract a method here
    mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
    mLastWorkerInfos = workerInfos;
}

// Return true means should update and mutual exclusiveness is guaranteed
private boolean shouldRebuildHashRingExclusively() {
      long lastUpdateTs = mLastUpdatedTimestamp.get();
      long currentTs = System.nanoTime();
      if (currentTs - lastUpdateTs <= mWorkerInfoUpdateIntervalNs) {
        return false;
      }
      return mLastUpdatedTimestamp.compareAndSet(lastUpdateTs, currentTs));
}
```
Of course your comments on thread safety mechanisms are still required.

Reply:
I feel that introducing another method will only add noise to the logic.

Reply:
extracted into a method